File: Sliding window/LET_no340_Longest Substring with At Most K Distinct Characters.py
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        if k == 0:
            return 0
        
        memo = collections.defaultdict(int)
        left = res = 0
        for right, c in enumerate(s):
            memo[c] += 1
            while left < right and len(memo) > k:
                memo[s[left]] -= 1
                if memo[s[left]] == 0:
                    del memo[s[left]]
                left += 1
            res = max(res, right - left + 1)
        
        return res

File: Sliding window/LET_no340_Longest Substring with At Most K Distinct Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        if (k == 0)
            return 0;

        unordered_map<int, int> memo;
        int res = 0;
        
        for (int left = 0, right = 0; right < s.size(); right++) {
            memo[s[right]]++;
            while (left < right && memo.size() > k) {
                if (--memo[s[left]] == 0)
                    memo.erase(memo.find(s[left]));
                left++;
            }

            res = max(res, right - left + 1);
        }

        return res;
    }
};